core/src/ldml/ldml_transforms.cpp

```diff
@@ -6,64 +6,191 @@
+   * which is already in input. Example (regex example):
+   * - str = "xxyyzz";
+   * - s/zz$/z/ -> match=2, output='z'
+   * - s/z$/zz/ -> match=2, output='zz'
+   *      (but could contract to match=0, output='' as already present)
+   * - s/zz$/zw/ -> match=2, output='zw'
+   *      (but could contract to match=1, output='w')
+   */
+  size_t matched              = 0;
+  std::u16string updatedInput = input;  // modified input
+  for (auto group = transform_groups.begin(); group < transform_groups.end(); group++) {
+    // for each transform group
+    // break out once there's a match
+    // TODO-LDML: reorders
+    // Assume it's a non reorder group
+    for (auto transform = group->begin(); transform < group->end(); transform++) {
+      // TODO-LDML: non regex implementation
+      // is the match area too short?
+      size_t subMatched = transform->match(updatedInput);
+      if (subMatched == 0)
+        continue;                                                       // no match
 
-  // int longest_index            = -1;  // index to longest match in simple_transforms
-  size_t longest_length = 0;  // length of longest match
-  // auto i                       = 0;   // counter
-  // TODO-LDML: need to handle partial matches.
-  // bool partial       = false;      // true if the longest was a partial match
-  for (auto it = simple.begin(); it != simple.end(); it++ /*,i++*/) {
-    auto list = it->first;
-    // see if it matches
-    if (longest_length >= list.size()) {
-      continue;  // already have a longer match
-    }
-    if (list.size() > ctxt.size()) {
-      continue;  // not enough ctxt for a match (TODO-LDML: partial match.)
-    }
-    // Now, check for match
-    bool is_match = true;
-    for (size_t j = 0; j < list.size(); j++) {
-      const auto found  = ctxt.at(ctxt.size() - j - 1);
-      const auto expect = list.at(list.size() - j - 1);
-      if (found != expect) {
-        // mismatch, break
-        is_match = false;
-        break;
+      // get the updated sub output
+      std::u16string subOutput = transform->apply(updatedInput, subMatched);
+
+      // remove the matched part of the updatedInput
+      updatedInput.resize(updatedInput.length() - subMatched);
+      updatedInput.append(subOutput);  // which could be empty
+
+      if (subMatched > output.size()) {
+        // including first time through
+        // expand match by amount subMatched prior to output
+        matched += (subMatched - output.size());
+      } // else: didn't match prior to the existing output, so don't expand 'match'
+
+      // now update 'output'
+      if (subOutput.length() >= output.length()) {
+        output = subOutput; // simple: all output
+      } else {
+        if (subMatched <= output.length()) {
+          output.resize(output.length() - subMatched);
+          output.append(subOutput);
+        } else {
+          output = subOutput;
+        }
       }
```

Comment:
Hmm, we have two identical output - subOutput statements in this if/else conglomeration. Another mild refactor may be called for. This matches the existing logic you have:

```suggestion
      // now update 'output'
      if (subOutput.length() >= output.length() || subMatched > output.length()) {
        output = subOutput; // simple: all output
      } else {
        output.resize(output.length() - subMatched);
        output.append(subOutput);
      }
```

But, I am not sure I understand the logic here -- should we be comparing `subOutput.length()` to `output.length()` at all?  Is this change more correct?

```suggestion
      // now update 'output'
      if (subMatched >= output.length()) {
        output = subOutput; // simple: all output
      } else {
        output.resize(output.length() - subMatched);
        output.append(subOutput);
      }
```

Reply:
This is to resize the output buffer though.